File: R/tagtools/R/read_cats.R

#' Read a CATS data file and convert to .nc
#'
#' Read a .csv file with data from a CATS tag deployment, including associated metadata, and store the resulting data in a .nc file.
#' @param fname is the file name of the CATS CSV file including the complete
#' path name if the file is not in the current working directory or in a
#' directory on the path. The .csv suffix is optional.
#' @param depid is a string containing the deployment identification code assigned
#' to this deployment, for example, 'mn12_186a'.
#' @return A string (constructed by: '\code{depid}_raw.nc'; for example, 'mn12_186a_raw.nc') containing the file name of the netCDF (.nc) file in which the output has been saved. This function
#' generates a netCDF file in the current working directory containing
#' 		the tag data variables, including:
#' 		\itemize{
#' 		\item {A, Accelerometer data structure}
#' 		\item {M, Magnetometer data structure}
#' 		\item {T, Temperature sensor data structure}
#' 		\item {info	Information structure for the deployment}
#' 		}
#' @note CATS loggers can produce very large csv files which are slow to
#' process. This function is (somewhat) optimised for speed and memory use so will
#' tolerate large files. But processing could be slow.
#' @export
#' @examples \dontrun{
#' nc_filename <- read_cats("my_cats_file.csv", "my_cats_deplyment_name")
#' load_nc("my_cats_deployment_name_raw.nc")
#' }
read_cats <- function(fname, depid) {
  V <- read_cats_csv(fname)
  info <- list(
    depid = depid,
    data_source = fname,
    data_nfiles = "1",
    data_format = "csv",
    device_serial = NULL,
    device_make = "CATS",
    device_type = "Archival",
    device_model_name = NULL,
    device_model_version = NULL,
    device_url = NULL,
    dephist_device_tzone = "0",
    dephist_device_regset = "dd-mm-yyyy HH:MM:SS",
    dephist_device_datetime_start = as.character(V[1, 1])
  )

  # time stuff
  dT <- as.numeric(diff(V[, 1] - rep(V[1, 1], nrow(V))))
  md <- stats::median(dT)
  km <- abs(dT - md) < 0.5 * md
  if (sum(km) < 0.75 * length(dT)) {
    warning("Many gaps in sampling. Inferred sampling rate may be inaccurate.\n")
  }
  # inferred sampling rate in Hertz
  sampling_rate <- 1 / mean(dT[km])

  # check which sensors are present
  Sens <- c("Acc", "Mag", "Gyr", "Temp", "Depth", "Light")
  Sens_name <- c(
    "3 axis Accelerometer", "3 axis Magnetometer", "3 axis Gyroscope",
    "Temperature", "Pressure", "Light level"
  )
  ax <- c(3, 3, 3, 1, 1, 1)
  # note: GPS are not well dealt with yet!!
  sl <- list()
  SS <- vector("logical", length = length(Sens))

  for (k in c(1:length(Sens))) {
    cols <- grep(Sens[k], names(V))
    if (any(!is.na(V[, cols]))) {
      sl <- paste(sl, Sens_name[k], ",", sep = "")
      SS[k] <- TRUE
    }
  }

  info$sensors_list <- sl
  nc_file <- paste(depid, "_raw.nc", sep = "")


  #******************************
  # I THINK CATS ACCEL IS
  # NED axes and need to change to FRU for tag tools
  #******************************

  #*********************************
  # HELPER FUNCTION
  save_sens_struct <- function(X, depid, sampling_rate, fname, name, naxes) {
    nc_file <- paste(depid, "_raw.nc", sep = "")
    cols <- grep(name, names(X))
    if (length(cols) > 0) {
      if (length(cols) < naxes) {
        warning(sprintf(" Warning: %d axes of %s missing in data\n", naxes - length(k), name))
      }
      if (naxes > 1) {
        # make sure column indices are ordered x, then y, then z
        # this assumes x, y, z cols of same sensor are named such that
        # alphabetical sorting --> x, y, z order
        cols <- cols[order(names(X)[cols])]
      } else {
        cols <- cols[1]
      }
      if (grepl("gyr", name)) {
        scf <- 0.001 # gyroscope unit is mrad/s. Multiply by 0.001 to get rad/s
      } else {
        scf <- 1 # all other units are standard
      }
      # pull names off data and make it a matrix
      X <- as.matrix(X[, cols])
      dimnames(X) <- NULL
      S <- sens_struct(
        data = X,
        sampling_rate, depid = depid, type = name
      )
      S$history <- "read_cats"
      S$files <- fname
      if (grepl("light", name)) {
        S$unit <- "1"
        S$unit_name <- "counts"
        S$unit_label <- "counts"
      }

      add_nc(nc_file, S, name)
    }
  } # end of save_sens_struct
  #********************************************************

  # now add sensor data
  for (k in c(1:length(Sens))) {
    if (SS[k]) {
      cols <- grep(Sens[k], names(V))
      save_sens_struct(V[, cols], depid, sampling_rate,
        fname,
        name = Sens[k], naxes = ax[k]
      )
    }
  }
  add_nc(nc_file, info, "info")
  return(nc_file)
} # end of read_cats

File: matlab/tagiofuns/read_cats.m
function 	ncfile = read_cats(fname,depid)

% 		ncfile = read_cats(fname,depid)
%		Read data files from a CATS data Logger. This function
%     generates a netCDF file in the current working directory containing
%		the variables in the data file including:
%		A Accelerometer data structure
% 		M Magnetometer data structure
% 		T Temperature sensor data structure
%		info	Information structure for the deployment
%      
% 		Inputs:
%     fname is the file name of the CATS csv file including the complete 
%      path name if the file is not in the current working directory or in a
%      directory on the path. The .csv suffix is not needed.
% 		depid is a string containing the deployment identification code assigned
%      to this deployment, for example, mn12_186a.
%  
% 		Results:
%     ncfile if the name of the NetCDF file containing the tag data. It will
%      be 'depid'_raw.nc (e.g., mn12_186a_raw.nc).
%
%		Warning: CATS loggers can produce very large csv files which are slow to
%		process. This function is optimised for speed and memory use so will
%		tolerate large files. But processing could be slow.
%
% 		Example:
% 		fn = read_cats('cats_test_sample','mn16_209a')
%		load_nc(fn)
%		% The workspace should now contain variables A, M, G, T, P, L and info
%		% each of which is a structure.
%
%     Valid: Matlab, Octave
%     markjohnson@st-andrews.ac.uk
%     last modified: 22 July 2017

if nargin<2,
    help read_cats
    return
end

[V,HDR] = read_cats_csv(fname) ;
info.depid=depid;
info.data_source=fname;
info.data_nfiles=1;
info.data_format='csv';
info.device_serial=[];
info.device_make='CATS';
info.device_type='Archival';
info.device_model_name=[];
info.device_model_version=[];
info.device_url=[];
info.dephist_device_tzone=0;
info.dephist_device_regset='dd-mm-yyyy HH:MM:SS';
info.dephist_device_datetime_start=datestr(V(1,1));

dT = diff(V(:,1)-V(1,1));
md = median(dT) ;
km = find(abs(dT-md)<0.5*md) ;
if length(km)<0.75*length(dT),
	fprintf('Warning: Many gaps in sampling. Inferred sampling rate may be inaccurate\n') ;
end	
FS = 1/(mean(dT(km))*3600*24) ;		% inferred sampling rate in Hertz

% find out which sensors are present
Sens = {'Acc','Mag','Gyr','Temp','Depth','Light'} ;
% TODO: find out what to do about GPS and other sensor channels
sl = '' ; SS = zeros(length(Sens),1) ;
ax = [3,3,3,1,1,1] ;
Sens_name = {'3 axis Accelerometer','3 axis Magnetometer','3 axis Gyroscope',...
	'Temperature','Pressure','Light level'} ;

for k=1:length(Sens),
	if any(strncmpi(HDR,Sens{k},length(Sens{k}))),
		sl = [sl,',',Sens_name{k}] ;
		SS(k) = 1 ;
	end
end

info.sensors_list = sl(2:end) ;
ncfile = [depid '_raw'] ;
save_nc(ncfile,info) ;

for k=1:length(Sens),
	if SS(k),
		save_sens_struct(V,depid,HDR,FS,fname,Sens{k},ax(k)) ;
	end
end
return


function		save_sens_struct(X,depid,CH,FS,fname,name,naxes)
%
k = find(strncmpi(CH,name,length(name))) ;
if ~isempty(k),
	if length(k)<naxes,
		fprintf(' Warning: %d axes of %s missing in data\n',naxes-length(k),name) ;
	end
	if naxes>1,
		ax = strvcat(CH{k}) ;
		ax = ax(:,end)' ;
		[ax,I] = sort(abs(ax)) ;		% sort into order of X, Y and Z
		k = k(I) ;
	else
		k = k(1) ;
	end
	if strncmpi(name,'gyr',3),
		scf = 0.001 ;	% gyroscope unit is mrad/s. Multiply by 0.001 to get rad/s
	else
		scf = 1 ;		% all other units are standard
	end
	S = sens_struct(X(:,k)*scf,FS,depid,name) ;
	S.history = 'read_cats' ;
	S.files = fname ;
	if strncmpi(name,'light',5),
		S.unit = '1' ;
		S.unit_name = 'counts' ;
		S.unit_label = 'counts' ;
	end
	add_nc([depid '_raw'],S) ;
end
